src/clocksi_interactive_tx_coord_fsm.erl

```diff
@@ -108,92 +119,180 @@ stop(Pid) -> gen_fsm:sync_send_all_state_event(Pid,stop).
-                                false ->
-                                    NewUpdatedPartitions = [IndexNode|UpdatedPartitions],
-                                    {reply, ok, execute_op, SD0#state{updated_partitions=NewUpdatedPartitions}};
-                                true->
-                                    {reply, ok, execute_op, SD0}
-                            end;
-                        {error, Reason} ->
-                            {reply, {error, Reason}, abort, SD0, 0}
-                    end;
-                {error, Reason} ->
-                    {reply, {error, Reason}, abort, SD0, 0}
-            end
+            {Key, Type, Param}=Args,
+	    Preflist = ?LOG_UTIL:get_preflist_from_key(Key),
+	    IndexNode = hd(Preflist),
+	    WriteSet = case lists:keyfind(IndexNode, 1, Updated_partitions) of
+			   false ->
+			       [];
+			   {IndexNode, WS} ->
+			       WS
+		       end,
+	    case ?CLOCKSI_DOWNSTREAM:generate_downstream_op(Transaction, IndexNode, Key, Type, Param, WriteSet) of
+		{ok, DownstreamRecord} ->
+		    NewUpdatedPartitions = case WriteSet of
+					       [] ->
+						   [{IndexNode,[{Key,Type,DownstreamRecord}]}|Updated_partitions];
+					       _ ->
+						   lists:keyreplace(IndexNode,1,Updated_partitions,
+								    {IndexNode,[{Key,Type,DownstreamRecord}|WriteSet]})
+					   end,
+		    _Res = gen_fsm:reply(Sender, ok),
+		    TxId = Transaction#transaction.txn_id,
+		    LogRecord = #log_record{tx_id=TxId, op_type=update,
+					    op_payload={Key, Type, DownstreamRecord}},
+		    LogId = ?LOG_UTIL:get_logid_from_key(Key),
+		    [Node] = Preflist,
+		    case ?LOGGING_VNODE:append(Node,LogId,LogRecord) of
+			{ok, _} ->
+			    {next_state, execute_op,
+			     SD0#state{updated_partitions= NewUpdatedPartitions}};
+			_ ->
+			    {next_state, abort, SD0}
+		    end;
+		{error, Reason} ->
+		    {reply, {error, Reason}, abort, SD0, 0}
+	    end
     end.
 
-%% @doc a message from a client wanting to start committing the tx.
-%%      this state sends a prepare message to all updated partitions and goes
+
+%% @doc this state sends a prepare message to all updated partitions and goes
 %%      to the "receive_prepared"state.
-prepare(timeout, SD0=#state{transaction=Transaction, updated_partitions=UpdatedPartitions, from=From}) ->
-    case length(UpdatedPartitions) of
-        0 ->
-            SnapshotTime = Transaction#transaction.snapshot_time,
-            gen_fsm:reply(From, {ok, SnapshotTime}),
-            {next_state, committing, SD0#state{state=committing, commit_time=SnapshotTime}};
+prepare(timeout, SD0=#state{
+                        transaction = Transaction,
+                        updated_partitions=Updated_partitions, from=From}) ->
+    case length(Updated_partitions) of
+        0->
+            Snapshot_time=Transaction#transaction.snapshot_time,
+            _Res = gen_fsm:reply(From, {ok, Snapshot_time}),
```

Comment:
I put this here because in the erlang docs gen_fsm:reply returns true
http://www.erlang.org/doc/man/gen_fsm.html#reply-2

But in the version of erlang I am using (Erlang R16B02 (erts-5.10.3) [source] [64-bit]) it returns the following:
Res: {#Ref<0.0.0.11727>,ok}

So maybe it is best to do a case statement to match either of these?